woocommerce/woocommerce-gutenberg-products-block
# Interactivity API and Product Button

This PR adds the Interactivity API to the Product Button to improve the UX that the user has when add a product in the cart from the Product Grid. Check the video to see the differences:

| Before | After |
|--------|--------|
|<video src=https://github.com/woocommerce/woocommerce-blocks/assets/4463174/12ba96a4-f542-4357-bac8-3bd2e777e88a /> |<video src=https://github.com/woocommerce/woocommerce-blocks/assets/4463174/f8b16c63-49a7-4448-8553-3c81b8d85844 />| 
 

Thanks to Interactivity API, the Product Button:
- can show the quantity of the product in the cart.
- can update the counter syncing the cart store redux.
- doesn't load JQuery anymore.

There is a follow-up PR that adds E2E tests woocommerce/woocommerce-blocks#10036.

### Testing

#### Automated Tests

#### User Facing Testing



1. Ensure that you are using the Blockified Product Catalog Template. If not, enable it.
2. Open the Site Editor and add the Mini Cart Block on the header. Save.
3. Visit the Product Catalog (`/shop`).
4. Click the button to ensure the product is added to the cart.
5. Ensure that the transition Add to cart -> loading status -> 1 in the cart is smooth.
6. Click on the Mini Cart.
7. Edit the quantity of the product.
8. Ensure that the changes are reflected in the Product Button.
9. Via dev tools ensure that the button element has the class `added` when the product is in the cart, and the button element has the class `loading` while the product is added to the cart.


##### Test ProductButton with cache plugin
1. Install the plugin [WP Optimize](https://wordpress.org/plugins/wp-optimize/).
2. Ensure that your cart is empty.
3. Ensure that you have the Mini Cart block in the header.
4. Enable the cache.
5. Visit the page with the Products block.
6. Refresh the page.
7. Add a product.
8. Refresh the page.
9. Ensure that the counter is updated with a smoother animation.
10. Open the Mini Cart block and change the quantity of the product in the cart.
11. Ensure that the counter inside the button is updated with a smoother animation.
12. Disable the plugin.
13. Visit the page with the Products block.
14. Open the Mini Cart block and change the quantity of the product in the cart.
15. Ensure that the counter inside the button is updated with a smoother animation.
16. Ensure that there isn't any regression.



### WooCommerce Visibility




### Performance Impact



### Changelog

> Product Button block: Integrate Interactivity API to make the block dynamically.
